New admin sign: add support for API communication & add tests

MVrachev opened the pull request:
<!--- Thanks for taking the time to fill out this pull request! -->
<!--- Please fill in the fields below to submit a pull request. 
The more information provided, the better. -->


# Description
New admin sign: add support for API communication 

Add support to fetch pending roles for signing and send the result
back to the API.
Added tests for all cases.


<!--- Please reference the issue(s) this PR fixes. -->
Related to issue #533 


# Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality not to work as expected)


# Additional requirements
- [x] Tests have been added for the bug fix or new feature
- [x] Docs have been added for the bug fix or new feature


# Code of Conduct
By submitting this PR, you agree to follow our [Code of Conduct](CODE_OF_CONDUCT.rst).

- [x] I agree to follow this project's Code of Conduct

kairoaraujo commented on repository_service_tuf/cli/admin/sign.py:
```diff
@@ -24,13 +26,53 @@
 
 
 @metadata.command()  # type: ignore
-@click.argument("root_in", type=click.File("rb"))
-@click.argument("prev_root_in", type=click.File("rb"), required=False)
```
**Note:** We can add it in a follow-up PR as well

I think it's still valid to keep it for local signing.
But I have some suggestions: 
1. Rename `root_in` for `SIGNING_JSON_INPUT`
   This input should be the JSON data from `GET [/api/v1/metadata/sign]` as a file input. The reason is that in the future, the user can sign any offline metadata (i.e., custom delegation targets)
2. Rename `prev_root_in` to `CURRENT_TRUSTED_METADATA`, to be any current trusted metadata`

It makes explicit and even more clear adding to the guide.

MVrachev commented on repository_service_tuf/cli/admin/sign.py:
```diff
@@ -39,20 +81,34 @@
+    if len(non_trusted) > 1:
+        rolename = Prompt.ask(
+            "\nChoose a metadata to sign",
+            choices=[role for role in non_trusted],
```
We don't.
We have prepared the code when we support offline targets signing. This question will NOT be called if there is only root for signing.

MVrachev replied:
It won't be in an inconsistent state if we go into that case. 
Given that you don't like it I will remove it.

lukpueh replied:
This is not a matter of me liking it or not, the code simply doesn't make sense here.

If there are multiple roots to sign, then RSTUF is in an inconsistent state. And if there are non-root roles to sign, it will be a problem for the subsequent lines of code, which expect a root role.

lukpueh commented on repository_service_tuf/cli/admin/sign.py:
```diff
@@ -88,9 +144,22 @@ def sign(root_in, prev_root_in, save) -> None:
         payload = SignPayload(signature=signature.to_dict())
-        json.dump(asdict(payload), save, indent=2)
+        with open(save.name, "w") as save_file:
+            json.dump(asdict(payload), save_file, indent=2)
```
Interesting. IMO it's better to mute mypy in that case, than to re-open an open file.

MVrachev commented on repository_service_tuf/cli/admin/sign.py:
```diff
@@ -24,13 +26,53 @@
+
+    pending_roles: Dict[str, Dict[str, Any]] = response_data.get(
+        "metadata", {}
+    )
```
I think the checks could provide good information about possible errors. I don't see it as an important thing to optimize now given the pending work.

lukpueh replied:
We can optimize this later.

MVrachev commented on repository_service_tuf/cli/admin/sign.py:
```diff
@@ -24,13 +26,53 @@
+
+    if settings.get("SERVER") is None:
+        api_server = Prompt.ask("\n[cyan]API[/] URL address")
+        settings.SERVER = api_server
```
I am okay to drop the settings check, but there is something I am missing as a usage for `settings.SERVER`.
I see we set it in other commands as well such as `add artifacts`, `delete artifacts` and `import-artifacts`.
@kairoaraujo if we set it in one of the commands will it be setup for my next command?

kairoaraujo replied:
What is missing in `settings.SERVER` is the documentation for the user.
As a practical example, I utilize my `~/.rstuf.yml` with `SERVER` defined, which empowers me to effectively not fill the server or address by the prompt every time. 
(BTW, @MVrachev, your code works with it, I had to comment to test it)
```shell
                                                                 
❯ rstuf admin metadata sign


┏━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┓
┃                                Metadata Signing Tool                                ┃
┗━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━┛
╭─ Error ─────────────────────────────────────────────────────────────────────────────╮
│ {"detail":{"message":"No metadata pending signing available","error":"Requires      │
│ bootstrap started. State: None"}}                                                   │
╰─────────────────────────────────────────────────────────────────────────────────────╯
                                                                                       
```

                 
The problem here is that we focus on the code and we forgot to document it.
For example, if you setup in the `rstuf.yaml` 

Regarding arg and prompt, prompt is better as it is interactive.

lukpueh replied:
> Regarding arg and prompt, prompt is better as it is interactive

IMO interactivity isn't better per se, and we did have the discussion before, about which inputs should be provided via cli arg, and which via prompt. From what I remember ,you can argue both ways.

pro all prompt: consistent "wizard" ux for all inputs
pro cli args and prompt: inputs like server address, input file path and output file path differ are different types of  inputs compared to the rest (keys, threshold, signers etc.) The former can be seen as config parameters to even use the tool, and the latter are what the tool is all about.

I lean towards using cli args for the config inputs, because they are generally easier to develop/maintain and to use. Independently, it's a nice additional feature to support providing these args via config file.

kairoaraujo replied:
I think I missed/forgot this discussion.
We should document it in our `docs/devel/design.rst`

MVrachev replied:
I think I support @kairoaraujo more on this one as for regular users of rstuf CLI it makes sense to have it configured.
If we want to continue this discussion I don't mind, but I don't feel this pr is the place to do it.
I will add `settings.SERVER` as an option input for now as that's how we have proceeded until now.
If we agree on something else we need to backtrack, not only for this function but everywhere to be consistent.

MVrachev commented:
I addressed all comments and added support for local signing as suggested by @kairoaraujo.
Also, added documentation on what will happen when local signing is used.

MVrachev commented:
Code coverage fails because of the https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/581 pr.
Not sure why it fails here as the files I have touch have full coverage.

kairoaraujo commented:
> Code coverage fails because of the #581 pr. Not sure why it fails here as the files I have touched on have full coverage.

I think it disappears if you rebase on top of the main.

MVrachev commented:
Rebased on top of main and fixed UT tests.
Ready for another review.
The code coverage issue remains.
PS: We will merge this pr after we release: #589.

MVrachev commented:
Rebased on top of main.
The only real commit is [Update docs](https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/579/commits/1bd4a434a3dd61f7debeb3c21bf431e57e9c11ea) where I update docs for the new `rstuf admin sign -h`.

kairoaraujo reviewed (approved):
LGTM

lukpueh commented on repository_service_tuf/cli/admin/sign.py:
```diff
@@ -88,9 +167,25 @@ def sign(root_in, prev_root_in, save) -> None:
+
+        console.print("Saved result to 'sign-payload.json'")
+
+    if not signing_json_input_file:
```
This is not documented, and I'm not sure if it's intuitive. Shouldn't it be possible to also post a locally read and then signed payload to the API?

lukpueh reviewed (commented):
Sorry for being late to the party. This looks pretty good. I pointed out some docs mistakes, which must be fixed, and programming recommendations, which could be fixed, and I raised question about behavior. Curious what others think.

MVrachev commented:
TODO: remove prompt for server if settings.SERVER and api-server miss.

Think if we can use default value of setting.SERVER for api-server for admin top-level command

MVrachev commented on docs/source/guide/index.rst:
```diff
@@ -179,11 +179,19 @@ sign (``sign``)
     Usage: rstuf admin metadata sign [OPTIONS] ROOT_IN [PREV_ROOT_IN]
```
Good catch!

MVrachev commented:
@lukpueh I addressed your comments in https://github.com/repository-service-tuf/repository-service-tuf-cli/pull/600.
I would appreciate it if you have a look there. :)
